src/Microsoft.ML.Data/Evaluators/Metrics/ConfusionMatrix.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Data
+{
+    /// <summary>
+    /// Represents the <a href="https://en.wikipedia.org/wiki/Confusion_matrix">confusion matrix</a> of the classification results.
+    /// </summary>
+    public sealed class ConfusionMatrix
+    {
+        /// <summary>
+        /// The calculated value of <a href="https://en.wikipedia.org/wiki/Precision_and_recall#Precision">precision</a> for each class.
+        /// </summary>
+        public ImmutableArray<double> PerClassPrecision { get; }
+
+        /// <summary>
+        /// The calculated value of <a href="https://en.wikipedia.org/wiki/Precision_and_recall#Recall">recall</a> for each class.
+        /// </summary>
+        public ImmutableArray<double> PerClassRecall { get; }
+
+        /// <summary>
+        /// The confsion matrix counts for the combinations actual class/predicted class.
+        /// The actual classes are in the rows of the table, and the predicted classes in the columns.
+        /// </summary>
+        public ImmutableArray<ImmutableArray<double>> ConfusionTableCounts { get; }
+
+        /// <summary>
+        /// The indicators of the predicted classes.
+        /// It might be the classes names, or just indices of the predicted classes, if the name mapping is missing.
+        /// </summary>
+        public IReadOnlyList<ReadOnlyMemory<char>> PredictedClassesIndicators;
+
+        internal readonly bool Sampled;
+        internal readonly bool Binary;
+
+        private readonly IHost _host;
+
+        internal ConfusionMatrix(IHost host, double[] precision, double[] recall, double[][] confusionTableCounts,
```

Comment:
>precision [](start = 54, length = 9)

Might be nice to have some asserts here to make poor readers of the code (like me!) aware of the intent here -- I assume `precision` and `recall` have the same length, that `confusionTableCounts` is a jagged array with that same length along both dimensions, but these are just guesses based on my understanding of the setting, which many people coming to the code fresh would not have. And even I'm not sure.